### ISSUE TITLE
Redirect is not working on localhost

### ISSUE THREAD

**Issue by weinerm** [Infocaster/UrlTracker#54]
**Describe the bug**
_Using version: 9.0.1_
I will preface this with the understanding this may not be a bug, but perhaps an issue with my local environment. A test redirect does not seem to be working, but is instead leading to a 404 page.

I have tried this with AppendPortNumber configuration set to both true and false.

**To Reproduce**
1. Set a hostname for a node to include localhost:44335 (or random port number)
2. Create a redirect, specifying an "Old Url" (I chose /donate)
3. Specify a redirect url (I chose https://google.com)
4. 301 Redirect
5. Force Redirect
6. Submit
7. Navigate to the URL: https://localhost:44335/donate (I get a 404 error)

**Expected behavior**
https://localhost:44335/donate redirects to https://google.com

---
Related to: #40

**Comment by D-Inventor** [comment]
I've managed to reproduce this issue by configuring a domain on a non-root node. I haven't been able to analyse the process, so I'll get back on that later today to see at which point the URL Tracker gets lost.

**Comment by D-Inventor** [comment]
I'm scratching my head a bit with this one. Could you perhaps [enable logging on the URL Tracker](https://github.com/Infocaster/UrlTracker/wiki/Configuration#enablelogging) and check the umbraco logs for the request that should be redirected? If you can find all the logs from the UrlTracker namespace, perhaps it gives some insights to what it's doing.

On verbose loglevel, you should see the url tracker detecting the incoming url, how many matches it finds for each interceptor and if it has found a handler or not.

**Comment by weinerm** [comment]
Verbose logs are attached. Logs are limited to me running the app locally, then navigating to the page in question (/donate) that should be redirecting.
[UmbracoTraceLog.MIKE-PC2.20220629.txt](https://github.com/Infocaster/UrlTracker/files/9013405/UmbracoTraceLog.MIKE-PC2.20220629.txt)

**Comment by D-Inventor** [comment]
Thank you very much for taking the time to share the logs. It appears that logging on the url tracker is disabled though, as there are no lines in the log file from the url tracker 😰. If you could please verify that logging is enabled with the `LoggingEnabled` option in [the configuration](https://github.com/Infocaster/UrlTracker/wiki/Configuration#version-9--10) and try again, that would be awesome!

I noticed that there is a spelling error in the configuration, where in the details it is named 'enableLogging', whereas in the intro, it is named 'LoggingEnabled'. The correct property name is `LoggingEnabled` and it should be set to `true`.

**Comment by weinerm** [comment]
I see UrlTracker on lines 5 & 6. Still need me to rerun it?

**Comment by weinerm** [comment]
[appsettings.txt](https://github.com/Infocaster/UrlTracker/files/9014254/appsettings.txt)
[UmbracoTraceLog.MIKE-PC2.20220629.txt](https://github.com/Infocaster/UrlTracker/files/9014255/UmbracoTraceLog.MIKE-PC2.20220629.txt)

Reran it with similar results. Also attached my appsettings.

**Comment by D-Inventor** [comment]
It's remarkable. It shows that the migrations are applied, so the database is all fine, but there are no logs at all from the middleware. Your appsettings are perfect.
This may be a silly question, but are you sure that you use the middleware in startup? If the middleware is used and the logger is enabled, I'd expect to see at least a log saying: "detected incoming request". The logs show nothing at all.

**Comment by weinerm** [comment]
Is there a good way to tell? I have this in Startup.cs as per the install instructions.

```
app.UseUmbraco()
                .WithMiddleware(u =>
                {
                    u.UseBackOffice();
                    u.UseWebsite();
                    u.UseUrlTracker();
                })
                .WithEndpoints(u =>
                {
                    u.UseInstallerEndpoints();
                    u.UseBackOfficeEndpoints();
                    u.UseWebsiteEndpoints();
                });
```

It's vanilla otherwise.

**Comment by D-Inventor** [comment]
That looks just fine, all you need is the call to `.UseUrlTracker()`.

Do you perhaps have a file _appsettings.development.json_ that is overwriting your serilog settings? Or some other appsettings transform?

**Comment by weinerm** [comment]
Yes, I had copied the UrlTracker configs in there, but forgot the log settings. Substantially larger log file now, though still didn't see but a few lines for UrlTracker. Hopefully it's helpful!
[UmbracoTraceLog.MIKE-PC2.20220630.txt](https://github.com/Infocaster/UrlTracker/files/9019768/UmbracoTraceLog.MIKE-PC2.20220630.txt)

**Comment by D-Inventor** [comment]
Alright, so according to these logs, the migrations have been executed and the composers have been discovered.

None of the logs have a SourceContext starting with `UrlTracker.xxx`, so nothing from the URL Tracker has actually been logged. This means that either the middleware doesn't get executed or the logger doesn't work.

This is the invoke method in the middleware:
```csharp
public async Task InvokeAsync(HttpContext context)
{
    Url url = context.Request.GetUrl();
    _logger.LogRequestDetected(url.ToString());
    
    await DoInterceptAsync(context, url);
    await _eventAggregator.PublishAsync(new UrlTrackerHandled(context, url));
}
```

As you can see, it logs pretty much instantly when you enter the middleware. I don't see any errors in your logs, so the middleware hasn't been interrupted.

Can we perhaps do some tests by changing your Startup.cs file like this?
```csharp
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, Microsoft.Extensions.Logging.ILogger<Startup> microsoftLogger, UrlTracker.Core.Logging.ILogger<Startup> logger)
        {
            app.UseUmbraco()
                .WithMiddleware(u =>
                {
                    u.UseBackOffice();
                    u.UseWebsite();

                    u.AppBuilder.Use((httpContext, next) =>
                    {
                        logger.LogError("Testing the logger to make sure that it works");
                        logger.LogDebug("Testing if the loglevel debug is actually logged with the url tracker logger");
                        microsoftLogger.LogError("Using native logger that works for sure to see if we actually get here");
                        return next();
                    });
                    u.UseUrlTracker();
                })
                .WithEndpoints(u =>
                {
                    u.UseInstallerEndpoints();
                    u.UseBackOfficeEndpoints();
                    u.UseWebsiteEndpoints();
                });
        }
```

This to test 2 things: We'll see if the URL Tracker logger logs anything at all (At error level to see if it works at all and at debug to see if it just doesn't log at debug level) and we'll also see if this part of the middleware is actually executed, by logging with the native asp net logger.

**Comment by D-Inventor** [comment]
And for completeness, you could place a breakpoint inside the `u.AppBuilder.Use` part and use the Immediate window in visual studio to try out `logger.IsEnabled(LogLevel.Debug)` to see if the logger is really working or not.

**Comment by weinerm** [comment]
Here we go, several logged items now. The breakpoint also reaffirms that logging is enabled.
[UmbracoTraceLog.MIKE-PC2.20220630.txt](https://github.com/Infocaster/UrlTracker/files/9022533/UmbracoTraceLog.MIKE-PC2.20220630.txt)

**Comment by D-Inventor** [comment]
Ah! That looks a lot better!

So this is what the logs tell me:

- It has found 0 static matches
- It has found 0 regex matches
- It has found 0 no longer exist matches
- It attempts to match on culture `en-US` and rootnodeid `5070`
- It attempts to match static urls on "https://localhost:44335/donate/", "https://localhost:44335/donate", "/donate/", "/donate", "donate/" or "donate"

See lines 6325 to 6329 in your latest log file.

I do believe that either /donate or /donate/ should match your redirect. How about the culture and rootnodeid? Does the rootnodeid match the node that you selected for your redirect? Does the culture match the culture you selected?

**Comment by weinerm** [comment]
Alright, making progress.

So that was not the correct root node ID in question, which is one problem. I had multiple localhost hostnames specified (with different port numbers) for a variety of nodes (so I could test multi-site locally). So I assume it is being confused by that. Not sure if you consider that a bug or not.

As a temporary measure, I think I've removed all the extra hostname mappings. Running into this log item now:
`{"@t":"2022-06-30T19:29:43.4899734Z","@mt":"Client error handling aborted: {reason}","@l":"Debug","reason":"Response is not 404","EventId":{"Id":2008},"SourceContext":"UrlTracker.Web.Events.UrlTrackerHandledNotificationHandler","RequestId":"4000006a-000e-f200-b63f-84710c7967bb","RequestPath":"/donate","ProcessId":22576,"ProcessName":"iisexpress","ThreadId":200,"AppDomainId":1,"AppDomainAppId":"f0e1d00be81d83fc64d93a0883fe7b051df99ecf","MachineName":"MIKE-PC2","Log4NetLevel":"DEBUG","HttpRequestId":"6b238674-c85f-41c1-955a-195360cfbedc","HttpRequestNumber":198,"HttpSessionId":"4a5f118d-cf33-243c-de30-1f0c6a7af621"}`

In short, I believe the "Force redirect" functionality may not be working. A (blank) page does exist at /donate

I was able to confirm the raw functionality was working by specifying a url that doesn't need a force redirect and it redirected as expected.

**Comment by D-Inventor** [comment]
ey! that's a step forward!

This line in your log is from the next step in the process, which is client error registration. That is: the step that registers 404 not found results on your application.

What's more interesting are logs with the template: `"{source} found {resultCount} results"`, as this is where the actual matching happens. If one of them has a resultCount > 0, then a match has been found.

The next interesting log should have the template: `"Intercept cancelled: {reason}, {redirect}"`, `"Redirect request to: {targetUrl}"` or `"Response converted to 410"` as these logs actually tell what the url tracker has done with the matches.

Do you have any of those?

**Comment by D-Inventor** [comment]
I'm gonna say that getting confused by the different portnumbers is a bug. I've checked in the source and I see that the check on port number is missing entirely.

Pretty much: if Umbraco supports it, then the URL Tracker should support it as well.

Fortunately that's an easy fix

**Comment by weinerm** [comment]
Well further good news...
After re-running the app, it is redirecting as expected. I assume it was just a byproduct of the port issue.

Many thanks for your help with this!

**Comment by D-Inventor** [comment]
Thanks to you as well! I've learned a lot and you've helped me make the URL Tracker even better.

I already have an open issue related to the malfunctioning domain check, so I've related this issue to that one.
I'll post an update soon (likely somewhere today) when I have fixed it!

**Comment by D-Inventor** [comment]
[Version 9.0.2-beta0002](https://www.nuget.org/packages/UrlTracker/9.0.2-beta0002) should have the fix for the domain and port matching. Do you have a chance to give it a try?

**Comment by weinerm** [comment]
Much improved!

Can you see if you can reproduce this issue?:

- Have two nodes with hostnames specified, both localhost but different ports
- Create a redirect for Root Node A
- Test the redirect is working for only Root Node A (and not B)
- Edit the redirect, changing it to Root Node B
- See if the redirect triggers for both Root Node A and Root Node B (it should only work for B)

**Comment by D-Inventor** [comment]
I cannot reproduce this.

Did you create a permanent or temporary redirect? Keep in mind that permanent redirects are cached by your browser, that may explain the behaviour that you're experiencing

**Comment by weinerm** [comment]
This definitely looks like a browser cache issue. Should be good to close this out, thanks!
